File: ui/boxcalc.js
/*
 * Computes box dimensions by expanding from a selected point.
 */
const { Point, Rectangle, Origin } = require("./point");

var context;

var maxdx = 144 * 3;
var maxdy = 72;

var imageData;

function setBoxBounds(dx, dy) {
    maxdx = dx;
    maxdy = dy;
}

function setCanvasContext(ctx) {
    context = ctx;
    imageData = context.getImageData(0, 0,
        context.canvas.width, context.canvas.height);
}

function computeBoxAtPoint(point) {

    const p = new Point(Math.round(point.x), Math.round(point.y));

    if (p.x < 0 || p.y < 0
        || p.x >= context.canvas.width || p.y >= context.canvas.height) {
        return undefined;
    }

    // Find the bottom edge.
    let ymax = p.plus(advanceLine(p, p, new Point(0, 1), maxdy));

    // Find the right edge.
    let xmax = p.plus(advanceLine(p, ymax, new Point(1, 0), maxdx));

    // Find the left edge.
    let xmin = p.plus(advanceLine(p, ymax, new Point(-1, 0), maxdx));

    // Recalculate the left and right edge, based just on the bottom line. Use
    // If the bottom line distance is reasonable, use whichever values lead to a
    // smaller box.
    let ybot = ymax.plus(0, 2);
    let xbotmax = ymax.plus(advanceLine(ybot, ybot, new Point(1, 0), maxdx));
    let xbotmin = ymax.plus(advanceLine(ybot, ybot, new Point(-1, 0), maxdx));
    let xbotdiff = xbotmax.x - xbotmin.x;
    if (xbotdiff > 20) {
        if (xmax.x > xbotmax.x) { xmax = xbotmax; }
        if (xmin.x < xbotmin.x) { xmin = xbotmin; }
    }

    // Find the top edge.
    let ymin = p.plus(advanceLine(
        new Point(xmin.x, p.y), new Point(xmax.x, p.y),
        new Point(0, -1), maxdy)
    );

    // Convert back to absolute coordinates
    let res = [ xmin.x, ymin.y, xmax.x, ymax.y ];

    // Sanity checks
    if (res[0] < 0) { res[0] = 0; }
    if (res[1] < 0) { res[1] = 0; }
    if (res[2] >= context.canvas.width) { res[2] = context.canvas.width - 1; }
    if (res[3] >= context.canvas.height) { res[3] = context.canvas.height - 1; }

    const r = new Rectangle(...res);
    return r;
}

/*
 * Returns the color at the given point. The point is given relative to an
 * origin that is the user's specified point.
 */
function colorAt(p) {
    if (p.x < 0 || p.y < 0
        || p.x >= imageData.width || p.y >= imageData.height) {
        return [ -100, -100, -100, -100 ];
    }
    const index = (p.y * imageData.width + p.x) * 4;
    return imageData.data.subarray(index, index + 3);
}

/*
 * Tests if the colors of two given points match.
 */
function sameColor(p1, p2) {
    const c1 = colorAt(p1), c2 = colorAt(p2);
    //console.log("Color at " + p1 + ": " + c1 + "; color at " + p2 + ": " + c2);
    for (let i = 0; i < 3; i++) {
        if (Math.abs(c1[i] - c2[i]) > 10) {
            return false;
        }
    }
    return true;
}

/*
 * Starting from the given line, advance the line by delta until the line
 * fails to match the original line.
 *
 * Returns the total displacement found (delta * number of steps taken).
 */
function advanceLine(pMin, pMax, delta, maxSteps) {
    for (let i = 1; i <= maxSteps; i++) {
        let totalDelta = delta.times(i);
        // The edges must match exactly.
        if (!sameColor(pMin, pMin.plus(totalDelta))) {
            return delta.times(i - 1);
        }
        if (!sameColor(pMax, pMax.plus(totalDelta))) {
            return delta.times(i - 1);
        }

        // Otherwise we check for a threshold error.
        let errs = 0;
        let count = 0;
        for (let pIter = pMin; pIter !== null; pIter = pIter.nextToward(pMax)) {
            count += 1;
            if (!sameColor(pIter, pIter.plus(totalDelta))) { errs += 1; }
        }
        if (errs > 5 || errs * 1.0 / count > 0.1) { return delta.times(i - 1); }
    }
    return delta.times(maxSteps);
}

module.exports = {
    setCanvasContext,
    computeBoxAtPoint,
    setBoxBounds,
    sameColor,
};

File: ui/boxcalc.rb
require_relative 'models'

class BoxCalculator

  include Marking

  def initialize(app, maxdx = 72 * 6, maxdy = 72)
    @app = app
    @maxdx = maxdx
    @maxdy = maxdy
    @min_width = 20
    @max_color_diff = 10
    @max_errs = 5
    @max_err_frac = 0.1
  end

  def compute_box_at_point(x, y = nil)
    p = y.nil? ? x.round : Point.new(x.round, y.round)

    max_rect = surface_rect
    return nil unless max_rect.include?(p)

    # Bottom edge
    ymax = p + advance_line(p, p, Point.new(0, 1), @maxdy)

    # Right edge
    xmax = p + advance_line(p, ymax, Point.new(1, 0), @maxdx)

    # Left edge
    xmin = p + advance_line(p, ymax, Point.new(-1, 0), @maxdx)

    # Find the line immediately below the bottom edge (by 2px to be safe).
    # Calculate a left and right boundary based on this, and choose the more
    # reasonable boundary.
    ybot = ymax + Point.new(0, 2)
    if max_rect.include?(ybot)
      xbotmax = ymax + advance_line(ybot, ybot, Point.new(1, 0), @maxdx)
      xbotmin = ymax + advance_line(ybot, ybot, Point.new(-1, 0), @maxdx)
      xbotdiff = xbotmax.x - xbotmin.x
      if xbotdiff > @min_width
        xmax = xbotmax if xmax.x > xbotmax.x
        xmin = xbotmin if xmin.x < xbotmin.x
      end
    end

    # Top edge
    ymin = p + advance_line(
      Point.new(xmin.x, p.y), Point.new(xmax.x, p.y),
      Point.new(0, -1), @maxdy
    )

    return Rectangle.new(
      max_rect.constrain(Point.new(xmin.x, ymin.y)),
      max_rect.constrain(Point.new(xmax.x, ymax.y))
    )
  end

  def advance_line(pmin, pmax, increment, max_steps)
    last_delta = Origin
    1.upto(max_steps) do |i|
      delta = increment * i

      # The edges must match exactly
      break unless same_color(pmin, pmin + delta)
      break unless same_color(pmax, pmax + delta)

      # For other points, check for a threshold error
      errs, count = 0, 0
      pmin.upto(pmax) do |pt|
        count += 1
        errs += 1 unless same_color(pt, pt + delta)
      end
      break if errs > @max_errs
      break if errs.to_f / count > @max_err_frac
      last_delta = delta
    end
    return last_delta
  end

  def color_at(x, y = nil)
    x, y = x.x, x.y if y.nil?
    if x < 0 || y < 0 || x >= surface.width || y >= surface.height
      return [ -100, -100, -100, -100 ]
    end

    index = (y * surface.width + x) * 4
    return surface_data[index, 4].unpack("C*")
  end

  def same_color(p1, p2)
    color_at(p1).zip(color_at(p2)).each do |v1, v2|
      return false if (v1 - v2).abs > @max_color_diff
    end
    return true
  end

  def surface_data
    @app.surface_data
  end

  def surface
    @app.surface
  end

  def surface_rect
    Rectangle.new(Origin, Point.new(surface.width - 1, surface.height - 1))
  end

end
